### DISCUSSION TITLE
动态过滤器正则否定表达疑问

### DISCUSSION THREAD

**Discussion by icateav** [the1812/Bilibili-Evolved#4919] [提问 / Q&A]
对于123这个up，我希望仅保留包含 节奏榜、新卡 的相关视频和动态
目前是这样描写  /^((?!节奏榜|新卡).)*$/ up:123
结果仅过滤了一小部分动态，依旧有很多不相关动态保留，是这个表达式不准确吗？应该怎样修改才能达到预期效果呢

**Comment by the1812** [comment]
`/^(?=节奏榜|新卡).+$/ up:123`

可以随便找个正则测试的网站试试, 比如 https://regex101.com/

**Reply by icateav** [reply]
<details>
  <summary>截图</summary>![屏幕截图 2024-09-19 001245](https://github.com/user-attachments/assets/d50a7ff7-f03b-432a-bf13-c9e4a1567e73)</details>
在网站中尝试 `^.*(?=节奏榜|新卡).+$` 能匹配到对应词语，但是在动态中 `/^.*(?=节奏榜|新卡).+$/ up:123` 反而没有产生任何过滤

说实话我有点晕，以前我只需匹配一个词语，使用 `/^(?!红警)/ up:红警HBK08` 就可以将所有不含红警的动态过滤，两个词的匹配我摸索了很久也没办法生效

**Reply by the1812** [reply]
我之前看错了, 你是希望仅保留, 那应该还是用 `?!` 替代 `?=`
但是如果前后都有 `.*` 或者 `.+` 的话也是不行的, 你希望检测的这个词语会被包含进这些组里
![Snipaste_2024-10-05_23-00-40](https://github.com/user-attachments/assets/e8b6ca83-7627-4ee5-9560-f9ded2fbb1fb)

**Reply by icateav** [reply]
尝试借助ai的能力，尝试使用`^(?!.*(?:新卡|节奏榜)).*$`，结果在网站中测试确实能发挥作用，匹配到所有不含关键词的内容，但是在b站实际应用效果却还是不能正确过滤，依旧保留很多不含关键词的动态和视频，我也完全弄不懂究竟是哪里有问题，头大，只能继续求助了
<details><summary>ai回复原文</summary>
要匹配除了“新卡”和“节奏榜”以外的内容，可以使用负向先行断言（Negative Lookahead）。以下是一个正则表达式示例：

```regex
^(?!.*(?:新卡|节奏榜)).*$
```

解释：
- `^` 表示字符串的开始。
- `(?!...)` 是负向先行断言，表示接下来的内容不应该匹配括号内的模式。
- `.*` 表示任意数量的任意字符。
- `(?:新卡|节奏榜)` 是非捕获组，表示匹配“新卡”或“节奏榜”。
- `$` 表示字符串的结束。

这个正则表达式会匹配任何不包含“新卡”和“节奏榜”的字符串。

例如：
- 输入：`今天天气不错`
  - 匹配结果：`今天天气不错`
- 输入：`新卡发布`
  - 匹配结果：不匹配
- 输入：`节奏榜更新`
  - 匹配结果：不匹配
- 输入：`新卡和节奏榜都很重要`
  - 匹配结果：不匹配

如果你需要更具体的上下文或例子，请提供更多详细信息。

**Comment by icateav** [comment]
还是搞不懂怎么保留多个词语，有没有会的大佬救一下
